Stringify query params which are objects as JSON

horiaradu opened the pull request:
Stringify query params which are objects in a way in which empty ararys
are preserved instead of removed (default querystring implementation).

fix: https://github.com/strongloop/strong-remoting/issues/324

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

horiaradu commented:
Any updates on this one?

richardpringle commented:
ok to test

richardpringle commented:
none of the tests you added are passing on my machine

richardpringle commented:
@slnode test please

richardpringle commented:
@horiaradu, the change that was made in this PR only affects the behavior of a `GET` request, where the source of the argument has not been defined, ie the default `GET` request behavior. Shouldn't we put this change in for any argument sourced from the query string?

Also, what happens if `typeof accept.type === 'object'` and `typeof val === 'object'`? You're still coercing the object into a string. Is your argument that any argument coming from the query should be a string? I'm don't necessarily disagree, I just want to clarify.

horiaradu commented:
@richardpringle 

What I want to accomplish is to serialize objects as strings when they need to go into the query string.

I only affected the GET request since I encountered this bug when I connected to a loopback API (as described [here](https://github.com/strongloop/strong-remoting/issues/324)) and only GET requests were used. I don't have enough knowledge of strong-remoting to know how the other requests are affected, but I can gladly try to help out in order to make this more generic.

richardpringle commented:
We should definitely make sure that the change is uniform and that any object parsed from the query string is coerced properly. 

If you could test the expected behaviour is the same for multiple verbs (GET and POST for example), that would be ideal.

horiaradu commented:
I will replicate the tests for other verbs as well.

On Thu, Aug 25, 2016 at 8:21 PM, Richard Pringle <notifications@github.com>
wrote:

> We should definitely make sure that the change is uniform and that any
> object parsed from the query string is coerced properly.
>
> If you could test the expected behaviour is the same for multiple verbs
> (GET and POST for example), that would be ideal.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/strongloop/strong-remoting/pull/325#issuecomment-242471066>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC_QOApzltVLd6uM2vXbQncBV-8d1zoWks5qjc8zgaJpZM4JRpHd>
> .
>

thaiat commented:
any reason why this is not merged ?

gunjpan commented:
@thaiat 
>any reason why this is not merged ?

Looks like there's pending item on this PR. Also, has merge conflicts.

@horiaradu : Can you please advise if you were able to finish:
> I will replicate the tests for other verbs as well.

Please resolve merge conflicts as well.

horiaradu commented:
I will resolve the conflicts asap.

On Wed, Oct 19, 2016 at 9:27 PM, Gunjan Pandya <notifications@github.com>
wrote:

> @thaiat <https://github.com/thaiat>
>
> any reason why this is not merged ?
>
> Looks like there's pending item on this PR. Also, has merge conflicts.
>
> @horiaradu <https://github.com/horiaradu> : Can you please advise if you
> were able to finish:
>
> I will replicate the tests for other verbs as well.
>
> Please resolve merge conflicts as well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/strongloop/strong-remoting/pull/325#issuecomment-254899773>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC_QOLNjmKytFtxN-b4XZc764YaWwoYHks5q1mESgaJpZM4JRpHd>
> .
>

horiaradu commented:
@gunjpan I've updated the PR.

* merge conflicts resolved
* added a couple of specs for POST methods

Check it out and let me know if I have to fix other things.

gunjpan commented:
@slnode ok to test

thaiat commented:
this should be ported back to latest 2.x version of strong-remoting as well

horiaradu commented:
I can take care of that after this PR is merged.

thaiat commented:
@horiaradu i needed that in my code asap so i did it there in case that helps you #377

gunjpan commented:
Looks like CI is stuck. Restarting tests.

gunjpan commented:
@slnode test please

horiaradu commented:
I would want to include the functionality with the unit tests as well so I
will just cherry pick the commit.

On Mon, Oct 24, 2016 at 4:30 PM, Gunjan Pandya <notifications@github.com>
wrote:

> Looks like CI is stuck. Restarting tests.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/strongloop/strong-remoting/pull/325#issuecomment-255740703>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC_QOEjOel-j5hl1v240DDbN2ZsyU5Seks5q3LLzgaJpZM4JRpHd>
> .
>

horiaradu commented:
I see that this PR appeared on the same topic: https://github.com/strongloop/strong-remoting/pull/378

Can somebody take care of this PR? It's been here for a while now...

Amir-61 commented:
@horiaradu I see @gunjpan is taking care of reviewing this PR. However before that you need to resolve merge conflicts. If this PR gets landed we can close https://github.com/strongloop/strong-remoting/pull/378

Thanks!

horiaradu commented:
I don't see any conflicts...

Amir-61 commented:
>I don't see any conflicts...

Oh it says:
> This branch is out-of-date with the base branch

I believe you need to rebase it on top of master.

horiaradu commented:
done.

Amir-61 commented:
@gunjpan could you PTAL.

gunjpan commented on lib/http-invocation.js:
```diff
@@ -148,7 +148,13 @@ HttpInvocation.prototype._processArg = function(req, verb, query, accept) {
+        query[name] = JSON.stringify(val);
+      } else {
+        query[name] = val;
+      }
```
@horiaradu : This needs to be applied for `query` case in here as well:
https://github.com/strongloop/strong-remoting/blob/master/lib/http-invocation.js#L127-L132

Better solution would be to extract it in a common function.

gunjpan commented on test/http-invocation.test.js:
```diff
@@ -193,6 +193,182 @@ describe('HttpInvocation', function() {
+        return [createEndpoint()];
+      };
+      var inv = givenInvocation(method, { ctorArgs: [], args: [], baseUrl: 'http://base' });
+
```
line 199:203 could be extracted to a single function as I see them being repeated.

gunjpan commented:
@horiaradu : I found pull #378 more elegant and generalized. 

@bajtos : Thoughts?

horiaradu commented:
@gunjpan The original reason this PR was made was to handle the **filter** query parameter. There, you have a full object with nested values which can be null / empty array.

#378 only handles the first level params and will fail if the null / empty array appears at a deeper level. You would need a recursive object traversal solution. Plus it does not add any tests and I think the unit tests are critical for a library such as this.

I've made the changes you suggested.

bajtos commented:
> @horiaradu : I found pull #378 more concise and generalized, because `JSON.stringify()` would retain numbers as numbers: From node REPL: 
> ```Console
> console.log(JSON.stringify({a: 2}))
{"a":2}
```

> @bajtos : Thoughts? 

I think #378 goes in a wrong direction. What happens when the argument value is a string? I think we will end up with `?key="value"` in the query, which is definitely not wanted.

```
> JSON.stringify('text')
'"text"'
```

This patch looks good to me as it is now, I think the only remaining step is to squash the commits to a single one with a good commit message ([guidelines](http://loopback.io/doc/en/contrib/git-commit-messages.html)). I'll leave the final call to @gunjpan.

gunjpan commented:
@bajtos : I see, thank you for clarifying the case of `string`

@horiaradu : PTAL at @bajtos' comments above. I'll further review your edits today. Thank you.

horiaradu commented:
done

gunjpan commented on test/http-invocation.test.js:
```diff
@@ -193,6 +193,146 @@ describe('HttpInvocation', function() {
   });
+
+  describe('createRequest', function() {
+    it('should create a simple request', function() {
```
@horiaradu : Please change all the `it` cases to reflect **imperative mood** as described in the Code style guidelines [here](http://loopback.io/doc/en/contrib/style-guide.html#test-naming).

gunjpan commented on test/http-invocation.test.js:
```diff
@@ -193,6 +193,146 @@ describe('HttpInvocation', function() {
+  describe('createRequest', function() {
+    it('should create a simple request', function() {
+      var inv = givenInvocationForEndpoint(null, []);
+
```
Please remove extra line here and
group together `var` declarations, for example:
```javascript
 var a = 'Hello';
 var b = 'How are you';

 function printMessage(a, b){
   console.log(a + ', ' + b);
}

```
 
Same applies to other it blocks

horiaradu commented:
@gunjpan check it out.

I will backport this to 2.X and create a PR as soon as this is merged.

gunjpan commented:
@horiaradu : Landed. Please backport it against `2.x`, and ping me for review. Thank you :)
